File: src/AtCoder/abc286/d.py
N, X = map(int, input().split())

def main():
    AB = [tuple(map(int, input().split())) for _ in range(N)]
    dp = [False] * (X+1)
    dp[0] = True
    for price, count in AB:
        dpn = dp.copy()
        for x in range(X+1):
            if dp[x] is False:
                continue
            for mul in range(1, count+1):
                nx = x + price * mul
                if nx > X:
                    break
                dpn[nx] = True
        dp = dpn
    if dp[X]:
        print('Yes')
    else:
        print('No')

main()

File: src/AtCoder/abc286/d.rs
#![allow(unused_imports, non_snake_case)]
use std::collections::*;
use itertools::Itertools;
use proconio::input;
use proconio::marker::*;

fn main() {
    input! {
        N: usize,
        X: usize,
        AB: [(usize, usize); N],
    }
    let mut dp = vec![false; X + 1];
    dp[0] = true;
    for (price, count) in AB {
        let mut dpn = dp.to_vec();
        for x in 0..=X {
            if !dp[x] {
                continue;
            }
            for mul in 1..=count {
                let nx = x + price * mul;
                if nx > X {
                    break;
                }
                dpn[nx] = true;
            }
        }
        dp = dpn;
    }
    if dp[X] {
        println!("Yes");
    } else {
        println!("No");
    }
    
}
